fili-core/src/test/groovy/com/yahoo/bard/webservice/web/endpoints/JobsServletReactiveChainforResultsEndpointSpec.groovy

```diff
@@ -152,73 +181,119 @@ class JobsServletReactiveChainforResultsEndpointSpec extends Specification {
-        setup:
-        JobsApiRequest apiRequest1 = new JobsApiRequest(null, "2", "", "", null, uriInfo, jobPayloadBuilder, apiJobStore, mockPreResponseStore, broadcastChannel)
+    def "When multiple results-are-ready notifications are sent, we only receive the results we care about"() {
+        given: "A proper PreResponseStore that allows us to know when values appear in the store"
+        CountDownLatch getTicket1Latch = new CountDownLatch(1)
+        preResponseStore.addGetLatch("ticket1", getTicket1Latch)
+        CountDownLatch saveTicket1Latch = new CountDownLatch(1)
+        preResponseStore.addSaveLatch("ticket1", saveTicket1Latch)
+
+        and:
+        JobsApiRequest apiRequest = new JobsApiRequest(
+                null,
+                "never",
+                "",
+                "",
+                null,
+                uriInfo,
+                jobPayloadBuilder,
+                apiJobStore
+        )
 
-        when: "PreResponse is available in the PreResponsestore"
-        mockPreResponseStore.save("ticket4", PreResponseTestingUtils.buildPreResponse("2016-04-24")).subscribe()
+        and:
+        TestSubscriber resultsSubscriber = new TestSubscriber()
 
-        and: "We start the async chain"
-        mockJobServlet.getPreResponseObservable("ticket4", apiRequest1)
-        //The delay is to ensure that we get the notification after async timeout
-        Thread.sleep(1000)
+        and:
+        jobsServlet.getResults("ticket1", apiRequest.asyncAfter).subscribe(resultsSubscriber)
 
-        and: "We receive the notification after async timeout"
-        broadcastChannel.publish("ticket4")
+        and: "We wait for the first attempt to get results from the store to come up empty before we add fake results"
+        getTicket1Latch.await(30, TimeUnit.SECONDS)
+        PreResponse results = Mock(PreResponse)
+        preResponseStore.save("ticket1", results)
 
-        then: "then we go to the PreResponseStore exactly once to get the ticket"
-        1 * mockPreResponseStore.get(_) >> Observable.just(Mock(PreResponse))
+        and: "We wait for the results to be successfully stored before sending a ready notification"
+        saveTicket1Latch.await(30, TimeUnit.SECONDS)
+
+        when: "We send a bunch of notifications through the broadcast channel, including multiples of ticket1"
+        ["ticket2", "ticket1", "ticket3", "ticket1", "ticket4", "ticket1"].each { broadcastChannel.publish(it) }
+
+        then: "The subscriber only gets the results for ticket1"
+        resultsSubscriber.assertValue(results)
+
+        cleanup:
+        preResponseStore.clearGetLatches()
+        preResponseStore.clearSaveLatches()
     }
 
-    def "If the PreResponse is available in the PreResponseStore and the notification from broadcastChannel is received within the async timeout, we go to the PreResponsestore twice"() {
```

Comment:
Whether or not the results are available actually has zero impact on this test. This test is testing how notifications from the broadcastChannel affects processing.